pkg/controller/status/controller.go

```diff
@@ -0,0 +1,411 @@
+)
+
+type Reported struct {
+	LastReportTime metav1.Time `json:"lastReportTime"`
+}
+
+type Configurator interface {
+	Config() *config.Controller
+}
+
+// Controller is the type responsible for managing the statusMessage of the operator according to the statusMessage of the sources.
+// Sources come from different major parts of the codebase, for the purpose of communicating their statusMessage with the controller.
+type Controller struct {
+	name      string
+	namespace string
+
+	client     configv1client.ConfigV1Interface
+	coreClient corev1client.CoreV1Interface
+
+	statusCh     chan struct{}
+	configurator Configurator
+
+	sources  []controllerstatus.Interface
+	reported Reported
+	start    time.Time
+
+	lock sync.Mutex
+}
+
+// NewController creates a statusMessage controller, responsible for monitoring the operators statusMessage and updating its cluster statusMessage accordingly.
+func NewController(client configv1client.ConfigV1Interface,
+	coreClient corev1client.CoreV1Interface,
+	configurator Configurator, namespace string) *Controller {
+	c := &Controller{
+		name:         "insights",
+		client:       client,
+		coreClient:   coreClient,
+		statusCh:     make(chan struct{}, 1),
+		configurator: configurator,
+		namespace:    namespace,
+	}
+	return c
+}
+
+func (c *Controller) triggerStatusUpdate() {
+	select {
+	case c.statusCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *Controller) controllerStartTime() time.Time {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.start.IsZero() {
+		c.start = time.Now()
+	}
+	return c.start
+}
+
+// LastReportedTime provides the last reported time in a thread-safe way.
+func (c *Controller) LastReportedTime() time.Time {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.reported.LastReportTime.Time
+}
+
+// SetLastReportedTime sets the last reported time in a thread-safe way.
+func (c *Controller) SetLastReportedTime(at time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.reported.LastReportTime.IsZero() {
+		klog.V(2).Infof("Initializing last reported time to %s", at.UTC().Format(time.RFC3339))
+	}
+	c.reported.LastReportTime.Time = at
+	c.triggerStatusUpdate()
+}
+
+// AddSources adds sources in a thread-safe way.
+// A source is used to monitor parts of the operator.
+func (c *Controller) AddSources(sources ...controllerstatus.Interface) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.sources = append(c.sources, sources...)
+}
+
+// Sources provides the sources in a thread-safe way.
+// A source is used to monitor parts of the operator.
+func (c *Controller) Sources() []controllerstatus.Interface {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.sources
+}
+
+func (c *Controller) merge(clusterOperator *configv1.ClusterOperator) *configv1.ClusterOperator {
+	// prime the object if it does not exist
+	if clusterOperator == nil {
+		clusterOperator = &configv1.ClusterOperator{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: c.name,
+			},
+		}
+	}
+
+	ctrlStatus := newControllerStatus()
```

Comment:
This means that every time when the `merge` is called then there's a new map created right? It doesn't seem right to me and I am not sure it will work correctly and I am also affraid of some leak.

Reply:
Shouldn't the GC take care of it? I think so... I didn't get any problem during my tests, but it wasn't a long run.
Maybe pass it to the struct would be a better approach. What do you think?

Reply:
I am not sure, but I would rather not rely on it. We would need to test it and probably profile it. Anyway it should be feasible to create the map once and only maintain its content.